### DISCUSSION TITLE
[실험] SOM-DST 논문 정리

### DISCUSSION THREAD

**Discussion by changwoomon** [bcaitech1/p3-dst-chatting-day#62] [Experiments]
# SOM-DST 논문 리뷰

## 문제 정의

Open-vocab 기반 DST 중, SOM-DST 가 속도와 성능에 있어 좋은 성능을 보이므로 적용해보기 위함

---
## 참고 자료
- 영상시청: [https://www.youtube.com/watch?v=nuclwoebdEM&ab_channel=KoreaUnivDSBA](https://www.youtube.com/watch?v=7Nwe2BBUZsw)
- 논문: [SOM-DST](https://arxiv.org/pdf/1911.03906.pdf)

---
### Ontology-based DST 문제점
- 실제 시나리오에 잘 대응하지 못함
- unseen value를 처리할 수 없음
- ontology가 많으면 처리 시간이 오래 걸림
### TRADE 문제점
- turn마다 slot의 모든 value를 생성해야해서 비효율적임

---
# SOM-DST

![image](https://user-images.githubusercontent.com/59998179/117854390-67382900-b2c4-11eb-9839-42066dd49b7a.png)

## Definition

<img src="https://render.githubusercontent.com/render/math?math=t"> : turn

<img src="https://render.githubusercontent.com/render/math?math=S^j">: slot

<img src="https://render.githubusercontent.com/render/math?math=V^j_t">: corresponding slot value

<img src="https://render.githubusercontent.com/render/math?math=J">: total number of such slots

<img src="https://render.githubusercontent.com/render/math?math=r^j_t\in O = \left\{\rm{CARRYOVER, DELETE, DONTCARE, UPDATE}\right\}">

<img src="https://render.githubusercontent.com/render/math?math=A_t">: System response

<img src="https://render.githubusercontent.com/render/math?math=U_t">: User utterance

---

## State Operation Predictor (Encoder)

Encoder 모델로 **pretrained BERT encoder** 사용

### Encoder Input을 만들기 위한 준비물

<img src="https://render.githubusercontent.com/render/math?math=D_t = A_t "> ⊕ ; ⊕ <img src="https://render.githubusercontent.com/render/math?math=U_t"> ⊕ <img src="https://render.githubusercontent.com/render/math?math=[SEP]">: dialogue utterances at turn t

- `;` <img src="https://render.githubusercontent.com/render/math?math=A_t">와 <img src="https://render.githubusercontent.com/render/math?math=U_t">를 구분하기 위한 스페셜 토큰
- `[SEP]` dialogue turn이 끝났다는 것을 표시하기 위한 스페셜 토큰

<img src="https://render.githubusercontent.com/render/math?math=B_t^j = [SLOT]^j"> ⊕ <img src="https://render.githubusercontent.com/render/math?math=S^j"> ⊕ - ⊕ <img src="https://render.githubusercontent.com/render/math?math=V_t^j"> : representation of the j-th slot-value pair

- j-th slot-value pair를 하나의 벡터로 aggregate
- <img src="https://render.githubusercontent.com/render/math?math=[SLOT]^j">

    [SLOT] 이라는 스페셜 토큰을 사용

    BERT의 [CLS] 토큰과 같은 역할

- <img src="https://render.githubusercontent.com/render/math?math=V_t^j">

    ![image](https://user-images.githubusercontent.com/59998179/117856211-4ec90e00-b2c6-11eb-8a89-952437c16dc9.png)

<img src="https://render.githubusercontent.com/render/math?math=B_t = B_t^1"> ⊕ ... ⊕ <img src="https://render.githubusercontent.com/render/math?math=B_t^J">: representation of the dialogue state at turn t

### Encoder Input

<img src="https://render.githubusercontent.com/render/math?math=X_t = [CLS]"> ⊕ <img src="https://render.githubusercontent.com/render/math?math=D_{t-1}"> ⊕ <img src="https://render.githubusercontent.com/render/math?math=D_t"> ⊕ <img src="https://render.githubusercontent.com/render/math?math=B_{t-1}">

segment id:            0            1         1

⇒ **Input : Sum(<img src="https://render.githubusercontent.com/render/math?math=X_t"> embedding, segment id embedding, positional embedding)**

dialogue history로 이전 턴의 dialogue utterances <img src="https://render.githubusercontent.com/render/math?math=D_{t-1}">을 사용한다.
dialogue history의 size: 1

모델이 입력으로 들어오는 dialogue 간의 Markov property를 가정
이전 turn dialogue state <img src="https://render.githubusercontent.com/render/math?math=B_{t-1}">은 전체 dialogue history를 압축적으로 표현하는 역할

### Encoder Output

<img src="https://render.githubusercontent.com/render/math?math=H_t \in \mathbb R^{\left|X_t\right|\times d}">: <img src="https://render.githubusercontent.com/render/math?math=h_t^{\rm X}"> (t=1...t) 까지 집합

<img src="https://render.githubusercontent.com/render/math?math=h_t^{[CLS]}, h_t^{[SLOT]^j} \in \mathrm R^d">

- <img src="https://render.githubusercontent.com/render/math?math=[CLS], [SLOT]^j">에 대응하는 output

<img src="https://render.githubusercontent.com/render/math?math=h_t^X = tanh(W_{pool} h_t^{[cls]})">

- <img src="https://render.githubusercontent.com/render/math?math=h_t^X">: 전체 input <img src="https://render.githubusercontent.com/render/math?math=X_t">를 sequence로 aggregate
- <img src="https://render.githubusercontent.com/render/math?math=W_{pool} \in \mathbb R^{d\times d}">: feed-forward layer with a learnable parameter

### State Operation Prediction

<img src="https://render.githubusercontent.com/render/math?math=P^j_{opr, t} = softmax(W_{opr} h_t^{[SLOT]^j})">

- <img src="https://render.githubusercontent.com/render/math?math=W_{opr} \in \mathbb R^{\left| O\right|\times d}"> : learnable parameter
- <img src="https://render.githubusercontent.com/render/math?math=P_{opr, t}^j \in \mathbb R^{\left| O\right|}"> : j-th slot의 turn t에서의 연산에 대한 확률 분포
- SOM-DST에서는 <img src="https://render.githubusercontent.com/render/math?math=\left| O\right| = 4">,

    <img src="https://render.githubusercontent.com/render/math?math=O = \left\{\rm{CARRYOVER, DELETE, DONTCARE, UPDATE}\right\}">

→ <img src="https://render.githubusercontent.com/render/math?math=r_t^j = argmax(P_{opr, t}^j)">

→ slot의 Operation의 결과가 `UPDATE` 일 때 slot value를 generation

- Encoder에서 나온 Operation의 결과가 `Update`인 경우를 집합으로 표현하면

    <img src="https://render.githubusercontent.com/render/math?math=\mathbb{U}_t = \left\{j|r_t^j = \rm{UPDATE}\right\}">, and its size as <img src="https://render.githubusercontent.com/render/math?math=J_t^\prime = \left| \mathbb{U}_t\right|">

    Recab for V

    ![image](https://user-images.githubusercontent.com/59998179/117856211-4ec90e00-b2c6-11eb-8a89-952437c16dc9.png)

---

## Slot Value Generator (Decoder)

- Encoder에서 나온 Operation의 결과가 `Update` 인 경우 해당 slot의 value를 예측
- SOM-DST의 generator는 value를 <img src="https://render.githubusercontent.com/render/math?math=J">가 아닌 <img src="https://render.githubusercontent.com/render/math?math=J^\prime_t">개의 slot에 대해서만 만들어준다.

    대부분의 경우에서 <img src="https://render.githubusercontent.com/render/math?math=J^\prime_t \ll J">이기 때문에 더 효율적이라고 주장

- Decoder 모델로 **GRU** 사용
    - 입력으로 word embedding vector <img src="https://render.githubusercontent.com/render/math?math=e_t^{j,k}">를 받으면서 **GRU**의 hidden state vector <img src="https://render.githubusercontent.com/render/math?math=g_t^{j, k}">를  recurrent하게 업데이트
    - <img src="https://render.githubusercontent.com/render/math?math=g_t^{j, 0} = h_t^{\rm x}">, <img src="https://render.githubusercontent.com/render/math?math=e_t^{j,0} = h_t^{[slot]j}">: GRU에 들어가는 초기값
    - <img src="https://render.githubusercontent.com/render/math?math=g_t^{j, k} = GRU(g_t^{j, k-1}, e_t^{j,k})">
    - <img src="https://render.githubusercontent.com/render/math?math=e_t^{j,k}">가 [EOS] 토큰이 나올때까지 진행
    - hidden state <img src="https://render.githubusercontent.com/render/math?math=g_t^{j, k}">는 k-th decoding step을 거치면서 vocabulary 와 user utterance의 단어에 대한 확률 분포로 변함

        <img src="https://render.githubusercontent.com/render/math?math=P^{j, k}_{vcb, t} = softmax(Eg^{j, k}_t) \in \mathbb R^{d_{vcb}}">

        - <img src="https://render.githubusercontent.com/render/math?math=E \in \mathbb R^{d_{vcb}\times d}">: Encoder와 Decoder가 서로 공유하는 word embedding matrix
            - <img src="https://render.githubusercontent.com/render/math?math=d_{vcb}">: vocabulary size

        <img src="https://render.githubusercontent.com/render/math?math=P^{j, k}_{ctx, t} = softmax(H_t g_t^{j, k}) \in \mathbb R^{\left|X_t\right|}">

        - user utterance의 단어에 대한 확률 분포

        <img src="https://render.githubusercontent.com/render/math?math=P^{j, k}_{val, t} = \alpha P^{j, k}_{vcb, t} + (1-\alpha) P^{j, k}_{ctx, t}"> : final output distribution

        - ![image](https://user-images.githubusercontent.com/59998179/117859561-20e5c880-b2ca-11eb-9194-81491f17d840.png)
          - <img src="https://render.githubusercontent.com/render/math?math=W_1 \in \mathbb R^{1\times (3d)}">: learnable parameter
          - <img src="https://render.githubusercontent.com/render/math?math=c^{j, k}_t = P^{j, k}_{ctx, t} H_t \in \mathbb R^d">: context vector

---

### State operation predictor

**Main Task**

state operation classification

**Auxiliary Task**

domain classification

state operation classification 외에도 domain classification을 보조 task로 사용하여 모델이 dialogue turn 간의 slot operation과 domain transition의 상관 관계를 학습하도록 함

<img src="https://render.githubusercontent.com/render/math?math=P_{dom, t} = softmax(W_{dom} h_t^{\rm X})">

- <img src="https://render.githubusercontent.com/render/math?math=W_{dom} \in \mathbb R^{d_{dom}\times d}">: learnable parameter
- <img src="https://render.githubusercontent.com/render/math?math=P_{dom, t} \in \mathbb R^{d_{dom}}">: turn t에서 domain에 대한 확률 분포
    - <img src="https://render.githubusercontent.com/render/math?math=d_{dom}">: # of domains defined in the dataset

**Average of the negative log-likelihood**

<img src="https://render.githubusercontent.com/render/math?math=L_{opr, t} = -\frac{1}{J}\sum_{j=1}^{J}(Y_{opr, t}^j)^\top log(P^j_{opr, t})">

<img src="https://render.githubusercontent.com/render/math?math=L_{dom, t} = -(Y_{dom, t})^\top log(P_{dom, t})">

- <img src="https://render.githubusercontent.com/render/math?math=Y_{dom, t} \in \mathbb R^{d_{dom}}">: one-hot vector for the ground truth domain
- <img src="https://render.githubusercontent.com/render/math?math=Y^j_{opr, t} \in \mathbb R^{\left| O\right|}">: one-hot vector for the ground truth operation for the j-th slot

### Slot value generator

**Average of the negative log-likelihood**

<img src="https://render.githubusercontent.com/render/math?math=L_{svg, t} = -\frac{1}{\left|\mathbb U_t\right|}\sum_{j\in\mathbb U_t}^{}\left[\frac{1}{K^j_t}\sum_{k=1}^{K^j_t}(Y_{val, t}^{j, k})^{\top}log(P^{j, k}_{val, t})\right]">

- <img src="https://render.githubusercontent.com/render/math?math=K_t^j">: # of tokens of the ground truth value that needs to be generated for the j-th slot
- <img src="https://render.githubusercontent.com/render/math?math=Y_{val, t}^{j, k} \in \mathbb R^{d_{vcb}}">: one-hot vector for the ground truth token that needs to be generated for the j-th slot at the k-th decoding step

### Final Loss

to minimized <img src="https://render.githubusercontent.com/render/math?math=L_{joint, t} = L_{opr, t} + L_{dom, t} + L_{svg, t}">

---

### Datasets

MultiWOZ 2.0 and MultiWOZ 2.1

### Training

- Encoder : Bert-base-uncased
- Decoder : GRU
- Hidden size : 768
- Optimizer : BertAdam
- Encoder LR and warmup : 4e-5, 0.1
- Decoder LR and warmup : 1e-4, 0.1
- Batch size : 32
- Dropout : 0.1
- Word Dropout 적용, 0.1확률로 word 를 [UNK] 로 바꿈
- Input max length : 256
- Training Epoch : 30

---

# 결과

## Joint Goal Accuracy

![image](https://user-images.githubusercontent.com/59998179/117859748-5e4a5600-b2ca-11eb-8535-dc1c7c185de1.png)

## Domain-specific Accuracy

![image](https://user-images.githubusercontent.com/59998179/117859798-6a361800-b2ca-11eb-8361-1c776b4d94bb.png)

## Latency

![image](https://user-images.githubusercontent.com/59998179/117859838-7ae68e00-b2ca-11eb-9147-23786fad7da2.png)

---

# 평가

- JGA, Domain-specific Accuracy 에서 SOTA 혹은 비슷한 수준의 성능을 보여줌
- inferecnce 타임이 매우 짧음에도 불구하고 좋은 성능을 보여줌
